### ISSUE TITLE
Trying to acquire a shared dataset lock may cause OverlappingFileLockException

### ISSUE THREAD

**Issue by RKrahl** [icatproject/ids.server#164]
The `class LockManager` manages locks that can be acquired on datasets. It features shared and exclusive locks. Multiple shared locks may be held on a dataset at the same time by different threads.

If a lock is requested, the `lock()` method is called on the `mainStorage` plugin. A typical implementation of that plugin call would try to acquire a file system lock (e.g. do something like calling `java.nio.channels.FileChannel.tryLock​(long position, long size, boolean shared)`), which essentially allows to extend the locking mechanism to external processes.

Now, there is a fundamental design flaw in the `LockManager` implementation: `mainStorage.lock()` is called for each individual lock request. E.g. it is called multiple times for the same dataset, if multiple shared locks are requested. With an implementation of the call in the plugin as sketched above, this will lead to attempting multiple overlapping file system locks, which will cause a `OverlappingFileLockException` to be thrown. As a result, the corresponding thread dies. If this thread is the FSM, the execution of deferred operations stops.

A correct implementation would call `mainStorage.lock()` only once, for the first lock request, keep the returned `AutoCloseable` and close it when the lock count drops to zero.

**Comment by RKrahl** [comment]
I contributed a [script](https://github.com/icatproject-contrib/scripts/blob/master/scripts/ids-lockmanager-bug164.py) that allows you to check whether your ids.server is susceptible to the issue.
